server/src-lib/Hasura/GraphQL/Parser/Schema.hs

```diff
@@ -0,0 +1,800 @@
+     how to parse their selection sets) so that it can parse the fragments.
+
+This presents some complications, since we need to build this information in a
+circular fashion. Currently, we do this in a very naïve way:
+
+  * We require selectionSetObject to specify the interfaces it implements /and/
+    require selectionSetInterface to specify the objects that implement it.
+
+  * We take advantage of our existing memoization mechanism to do the knot-tying
+    for us (see Note [Tying the knot] in Hasura.GraphQL.Parser.Class).
+
+You may notice that this makes it possible for the definitions to be
+inconsistent: we could construct an interface parser that parses some object
+type, but forget to specify that the object type implements the interface. This
+inconsistency is currently completely unchecked, which is quite unfortunate. It
+also means we don’t support remote schema-defined object types that implement
+interfaces we generate, since we don’t know anything about those types when we
+construct the interface.
+
+Since we don’t make very much use of interface types at the time of this
+writing, this isn’t much of a problem in practice. But if that changes, it would
+be worth implementing a more sophisticated solution that can gather up all the
+different sources of information and make sure they’re consistent. -}
+
+data InputObjectInfo = InputObjectInfo ~[Definition InputFieldInfo]
+-- Note that we can't check for equality of the fields since there may be
+-- circularity. So we rather check for equality of names.
+instance Eq InputObjectInfo where
+  InputObjectInfo fields1 == InputObjectInfo fields2
+    =  Set.fromList (fmap dName fields1)     == Set.fromList (fmap dName fields2)
+
+data ObjectInfo = ObjectInfo
+  { oiFields :: ~[Definition FieldInfo]
+    -- ^ The fields that this object adds on top of the fields of the interfaces
+    -- that it implements.
+  , oiImplements :: ~[Definition InterfaceInfo]
+    -- ^ The interfaces that this object implements (inheriting all their
+    -- fields). See Note [The interfaces story] for more details.
+  }
+-- Note that we can't check for equality of the fields and the interfaces since
+-- there may be circularity. So we rather check for equality of names.
+instance Eq ObjectInfo where
+  ObjectInfo fields1 interfaces1 == ObjectInfo fields2 interfaces2
+    =  Set.fromList (fmap dName fields1    ) == Set.fromList (fmap dName fields2    )
+    && Set.fromList (fmap dName interfaces1) == Set.fromList (fmap dName interfaces2)
+
+-- | Type information for a GraphQL interface; see Note [The interfaces story]
+-- for more details.
+--
+-- Note: in the current working draft of the GraphQL specification (> June
+-- 2018), interfaces may implement other interfaces, but we currently don't
+-- support this.
+data InterfaceInfo = InterfaceInfo
+  { iiFields :: ~[Definition FieldInfo]
+    -- ^ Fields declared by this interface. Every object implementing this
+    -- interface must include those fields.
```

Comment:
I was just about to write a comment here that I wasn't sure about the word "include" here. I thought I had written the code that uses `ObjectInfo`s to assume that the fields of the object is the union of its `oiFields` *plus* the `iiFields` of all its `oiImplements`. However, when I set out to check this claim, it turns out that `selectionSetObject` seemingly only parses the fields corresponding to its `oiFields`, suggesting that `oiFields` already includes the fields specified by all of its interfaces. And the introspection code says that an object only has, as its fields, the ones specified in its `oiFields`. So I must have written part of this code while being partially confused about the intention of `oiFields`: should it be _only_ the fields _in addition_ to the required ones (my original plan), or the fields _including_ those it's required to have according to its `oiImplements`? I'll have to go through the code to check that the implementation is consistent. @lexi-lambda, do you think one design is perhaps preferable over the other?

Reply:
I waffled a little on the wording here, too, for some of the same reasons, but I didn’t realize all those details.

The GraphQL perspective here is clearly that an object type “owns” all of its fields, and it simply must include all the fields required by an interface to implement that interface. However, that perspective makes illegal states trivially representable. On the other hand, the inheritance approach doesn’t *really* buy us any additional correctness, since we wouldn’t know how to parse the extra “inherited” fields, anyway.

So I propose we keep the code the same, but change the comment on `oiFields` to document the invariant.

Reply:
In fact, perhaps we can't really avoid specifying an object's fields in full detail, as the type of a given object field may be different from that same field in an interface that it implements: [the GraphQL spec](http://spec.graphql.org/June2018/#sec-Objects) (scroll down to "Type Validation") doesn't say that the type of the field of the object must _equal_ the type of the field of the interface, but merely that it must be a _subtype_ (where this is defined in the GraphQL spec).

(And I should've realized this earlier as I partially implemented this subtype checking algorithm in `Remote.hs`.)

Reply:
I [pushed a commit](https://github.com/hasura/graphql-engine/pull/4111/commits/b5a60af52d174d37ccb29621cf395f246c9f301e) fixing the comments for `selectionSetObject` and `ObjectInfo`.